src/server/ua_server_ns0.c

```diff
@@ -558,7 +558,7 @@ UA_Server_initNS0(UA_Server *server) {
     /* LocaleIdArray */
     UA_LocaleId locale_en = UA_STRING("en");
     retVal |= writeNs0VariableArray(server, UA_NS0ID_SERVER_SERVERCAPABILITIES_LOCALEIDARRAY,
-                                    &locale_en, 1, &UA_TYPES[UA_TYPES_LOCALEID]);
+                                    &locale_en, 1, &UA_TYPES[UA_TYPES_STRING]);
```

Comment:
Shouldn't it still be LOCALEID?
If that's not correctly shown on the client side (e.g. it only shows ExtensionObject) then we maybe have a problem in encoding which we should try to fix?

Reply:
in the generated code, it is defined as a UA_TYPE_STRING :

```
/* LocaleIdArray - ns=0;i=2271 */
{
UA_VariableAttributes attr = UA_VariableAttributes_default;
attr.minimumSamplingInterval = 0.000000;
attr.userAccessLevel = 1;
attr.accessLevel = 1;
attr.valueRank = 1;
attr.arrayDimensionsSize = 1;
attr.arrayDimensions = (UA_UInt32 *)UA_Array_new(1, &UA_TYPES[UA_TYPES_UINT32]);
attr.arrayDimensions[0] = 0;
attr.dataType = UA_NODEID_NUMERIC(ns0,12);
UA_String *variablenode_ns_0_i_2271_variant_DataContents = (UA_String*) UA_alloca(UA_TYPES[UA_TYPES_STRING].memSize * 1); 
UA_init(&variablenode_ns_0_i_2271_variant_DataContents[0], &UA_TYPES[UA_TYPES_STRING]);
UA_Variant_setArray( &attr.value, variablenode_ns_0_i_2271_variant_DataContents, (UA_Int32) 1, &UA_TYPES[UA_TYPES_STRING]);
attr.displayName = UA_LOCALIZEDTEXT("", "LocaleIdArray");
attr.description = UA_LOCALIZEDTEXT("", "A list of locales supported by the server.");
attr.writeMask = 0;
attr.userWriteMask = 0;
retVal |= UA_Server_addVariableNode(server,
UA_NODEID_NUMERIC(ns0,2271),
UA_NODEID_NUMERIC(ns0,0),
UA_NODEID_NUMERIC(ns0,0),
UA_QUALIFIEDNAME(ns0, "LocaleIdArray"),
UA_NODEID_NUMERIC(ns0,68),
attr,
NULL, NULL);
UA_Array_delete(attr.arrayDimensions, 1, &UA_TYPES[UA_TYPES_UINT32]);
}

retVal |= UA_Server_addReference(server, UA_NODEID_NUMERIC(ns0,2271), UA_NODEID_NUMERIC(ns0,46), UA_EXPANDEDNODEID_NUMERIC(ns0, 2268), false);
```

Reply:
According to the spec the LocaleIdArray should be of type `LocaleId[]` and not `String[]`

Reply:
Ok thanks.

Reply:
Do you have any news about this issue?

Reply:
In this case we have the same problem for all opaque type. Eg: CurrentTime sould be of type UtcTime but is currently as type DateTime.

![image](https://user-images.githubusercontent.com/13799456/33769119-31d760e8-dc29-11e7-84f5-038a552138a5.png)

and readCurrentTime use UA_TYPES_DATETIME instead of UtcTime

```
static UA_StatusCode
readCurrentTime(UA_Server *server, const UA_NodeId *sessionId, void *sessionContext,
                const UA_NodeId *nodeid, void *nodeContext, UA_Boolean sourceTimeStamp,
                const UA_NumericRange *range, UA_DataValue *value) {
    if(range) {
        value->hasStatus = true;
        value->status = UA_STATUSCODE_BADINDEXRANGEINVALID;
        return UA_STATUSCODE_GOOD;
    }
    UA_DateTime currentTime = UA_DateTime_now();
    UA_StatusCode retval = UA_Variant_setScalarCopy(&value->value, &currentTime,
                                                    &UA_TYPES[UA_TYPES_DATETIME]);
    if(retval != UA_STATUSCODE_GOOD)
        return retval;
    value->hasValue = true;
    if(sourceTimeStamp) {
        value->hasSourceTimestamp = true;
        value->sourceTimestamp = currentTime;
    }
    return UA_STATUSCODE_GOOD;
}
```

If I try to modify readCurrentTime (using UA_TYPES_UTCTIME) 
```
static UA_StatusCode
readCurrentTime(UA_Server *server, const UA_NodeId *sessionId, void *sessionContext,
                const UA_NodeId *nodeid, void *nodeContext, UA_Boolean sourceTimeStamp,
                const UA_NumericRange *range, UA_DataValue *value) {
    if(range) {
        value->hasStatus = true;
        value->status = UA_STATUSCODE_BADINDEXRANGEINVALID;
        return UA_STATUSCODE_GOOD;
    }
    UA_DateTime currentTime = UA_DateTime_now();
    UA_StatusCode retval = UA_Variant_setScalarCopy(&value->value, &currentTime,
                                                    &UA_TYPES[UA_TYPES_UTCTIME]);
    if(retval != UA_STATUSCODE_GOOD)
        return retval;
    value->hasValue = true;
    if(sourceTimeStamp) {
        value->hasSourceTimestamp = true;
        value->sourceTimestamp = currentTime;
    }
    return UA_STATUSCODE_GOOD;
}
```
UaExpert display extensionobjet for the value (like LocaleIdArray)

![image](https://user-images.githubusercontent.com/13799456/33769347-1f283dd6-dc2a-11e7-8d91-3c76c9b1073c.png)